File: examples/python/upload_and_publish.py
import json
import os
import sys

import requests


BASE_URL = os.getenv("BASE_URL", "https://api.voxelshelf.com/api")
API_KEY = os.getenv("VOXELSHELF_API_KEY")
UPLOAD_ID = os.getenv("UPLOAD_ID", "upl_replace_me")
ASSET_STORAGE_KEY = os.getenv("ASSET_STORAGE_KEY", "uploads/demo/model.zip")

if not API_KEY:
    print("Missing VOXELSHELF_API_KEY", file=sys.stderr)
    sys.exit(1)

headers = {
  "Authorization": f"Bearer {API_KEY}",
}

complete_resp = requests.post(
  f"{BASE_URL}/v1/uploads/complete",
  headers={
    **headers,
    "Content-Type": "application/json",
    "X-Request-Id": "req_py_upload_complete_001",
    "Idempotency-Key": "idem-py-upload-complete-001",
  },
  json={"uploadId": UPLOAD_ID},
  timeout=30,
)
complete_resp.raise_for_status()

create_payload = {
  "title": "Python Example Product",
  "description": "Created from Python API example.",
  "category": "Home & Decor",
  "subcategory": "Decor & Lighting",
  "printMaterials": ["PLA"],
  "supports": "Optional",
  "difficulty": "Easy",
  "isPresupported": False,
  "license": "Personal",
  "price": 9,
  "images": [{"url": "https://images.voxelshelf.com/demo-python.jpg", "alt": "Python demo"}],
  "files": [
    {
      "storageKey": ASSET_STORAGE_KEY,
      "filename": "model.zip",
      "size": 1234567,
      "mime": "application/zip",
      "kind": "Archive",
    }
  ],
}

create_resp = requests.post(
  f"{BASE_URL}/v1/creator/products",
  headers={
    **headers,
    "Content-Type": "application/json",
    "X-Request-Id": "req_py_product_create_001",
    "Idempotency-Key": "idem-py-product-create-001",
  },
  data=json.dumps(create_payload),
  timeout=30,
)
create_resp.raise_for_status()
draft = create_resp.json()
product_id = draft.get("id")
if not product_id:
  raise RuntimeError("create product response did not include id")

publish_resp = requests.post(
  f"{BASE_URL}/v1/creator/products/{product_id}/publish",
  headers={
    **headers,
    "X-Request-Id": "req_py_product_publish_001",
    "Idempotency-Key": "idem-py-product-publish-001",
  },
  timeout=30,
)
publish_resp.raise_for_status()
print(publish_resp.json())

File: examples/node/upload-and-publish.mjs
const baseUrl = process.env.BASE_URL ?? "https://api.voxelshelf.com/api";
const apiKey = process.env.VOXELSHELF_API_KEY;
const uploadId = process.env.UPLOAD_ID ?? "upl_replace_me";
const fileStorageKey = process.env.ASSET_STORAGE_KEY ?? "uploads/demo/model.zip";

if (!apiKey) {
  console.error("Missing VOXELSHELF_API_KEY");
  process.exit(1);
}

async function run() {
  const completeUploadResponse = await fetch(`${baseUrl}/v1/uploads/complete`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
      "X-Request-Id": "req_node_upload_complete_001",
      "Idempotency-Key": "idem-node-upload-complete-001",
    },
    body: JSON.stringify({ uploadId }),
  });

  if (!completeUploadResponse.ok) {
    const body = await completeUploadResponse.text();
    throw new Error(`complete failed (${completeUploadResponse.status}): ${body}`);
  }

  const createProductResponse = await fetch(`${baseUrl}/v1/creator/products`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
      "X-Request-Id": "req_node_product_create_001",
      "Idempotency-Key": "idem-node-product-create-001",
    },
    body: JSON.stringify({
      title: "Node Example Product",
      description: "Created from Node API example.",
      category: "Home & Decor",
      subcategory: "Decor & Lighting",
      printMaterials: ["PLA"],
      supports: "Optional",
      difficulty: "Easy",
      isPresupported: false,
      license: "Personal",
      price: 12,
      images: [{ url: "https://images.voxelshelf.com/demo-node.jpg", alt: "Node demo" }],
      files: [
        {
          storageKey: fileStorageKey,
          filename: "model.zip",
          size: 1234567,
          mime: "application/zip",
          kind: "Archive",
        },
      ],
    }),
  });

  if (!createProductResponse.ok) {
    const body = await createProductResponse.text();
    throw new Error(`create product failed (${createProductResponse.status}): ${body}`);
  }

  const draft = await createProductResponse.json();
  const productId = draft?.id;
  if (!productId) {
    throw new Error("create product response did not include id");
  }

  const publishResponse = await fetch(`${baseUrl}/v1/creator/products/${productId}/publish`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "X-Request-Id": "req_node_product_publish_001",
      "Idempotency-Key": "idem-node-product-publish-001",
    },
  });

  if (!publishResponse.ok) {
    const body = await publishResponse.text();
    throw new Error(`publish failed (${publishResponse.status}): ${body}`);
  }

  const published = await publishResponse.json();
  console.log("Published product:", published);
}

run().catch((error) => {
  console.error(error);
  process.exit(1);
});
